radixdlt/src/main/java/com/radixdlt/consensus/safety/SafetyRules.java

```diff
@@ -88,23 +89,23 @@ public void process(QuorumCertificate qc) {
 	 * @param highestCommittedQC highest known committed QC
 	 * @return signed proposal object for consensus
 	 */
-	public Proposal signProposal(Vertex proposedVertex, QuorumCertificate highestCommittedQC, long payload) {
+	public Proposal signProposal(UnverifiedVertex proposedVertex, QuorumCertificate highestCommittedQC, long payload) {
 		final Hash vertexHash = this.hasher.hash(proposedVertex);
 		ECDSASignature signature = this.signer.sign(vertexHash);
 		return new Proposal(proposedVertex, highestCommittedQC, this.self, signature, payload);
 	}
 
-	private static VoteData constructVoteData(Vertex proposedVertex, BFTHeader proposedHeader) {
-		final BFTHeader parent = proposedVertex.getQC().getProposed();
+	private static VoteData constructVoteData(VerifiedVertex proposedVertex, BFTHeader proposedHeader) {
+		final BFTHeader parent = proposedVertex.getParentHeader();
 
 		// Add a vertex to commit if creating a quorum for the proposed vertex would
 		// create three consecutive qcs.
 		final BFTHeader toCommit;
-		if (proposedVertex.getView().equals(proposedVertex.getParentHeader().getView().next())
-			&& !proposedVertex.getParentHeader().getView().isGenesis() && !proposedVertex.getGrandParentHeader().getView().isGenesis()
-			&& proposedVertex.getParentHeader().getView().equals(proposedVertex.getGrandParentHeader().getView().next())
+		if (!proposedVertex.touchesGenesis()
```

Comment:
Maybe this would be clearer with the logic inverted.  I had to read this a couple of time to convince myself it was correct.  This might be better:
```
if (proposedVertex.touchesGenesis()
	|| !proposedVertex.hasDirectParent()
	|| !proposedVertex.parentHasDirectParent()
) {
	toCommit = null;
} else {
	toCommit = proposedVertex.getGrandParentHeader();
}
```

It's certainly better than it was before in any case.